qiskit/quantum_info/operators/scalar_op.py

```diff
@@ -0,0 +1,264 @@
+        # update the coefficient and dimensions
+        if isinstance(other, ScalarOp):
+            if self.coeff is None:
+                coeff = other.coeff
+            elif other.coeff is None:
+                coeff = self.coeff
+            else:
+                coeff = self.coeff * other.coeff
+            return ScalarOp(input_dims, coeff=coeff)
+
+        # If we are composing on the full system we return the
+        # other operator with reshaped dimensions
+        if qargs is None:
+            ret = other.reshape(input_dims, output_dims)
+            if self.coeff is None or self.coeff == 1:
+                return ret
+            return self.coeff * ret
+        # Otherwise compose using other operators method
+        # Note that in this case that operator must know how to initalize
+        # from an ScalarOp either using its to_operator method or having
+        # a specific initialization case.
+        return other.__class__(self).compose(
+            other, qargs=qargs, front=not front)
+
+    def power(self, n):
+        """Return the compose of a operator with itself n times.
+
+        Args:
+            n (int): the number of times to compose with self (n>0).
+
+        Returns:
+            BaseOperator: the n-times composed operator.
+
+        Raises:
+            QiskitError: if the input and output dimensions of the operator
+            are not equal, or the power is not a positive integer.
+        """
+        if self.coeff is None:
+            return self
+        if n == 0:
+            # Raising to zero power returns identity
+            return ScalarOp(self._input_dims)
+        coeff = self.coeff ** n
+        return ScalarOp(self._input_dims, coeff=coeff)
+
+    def _add(self, other):
+        """Return the operator self + other.
+
+        Args:
+            other (BaseOperator): an operator object.
+
+        Returns:
+            ScalarOp: if other is an ScalarOp.
+            BaseOperator: if other is not an ScalarOp.
+
+        Raises:
+            QiskitError: if other has incompatible dimensions.
+        """
+        if not isinstance(other, BaseOperator):
+            other = Operator(other)
+        self._validate_add_dims(other)
+        if isinstance(other, ScalarOp):
+            coeff1 = 1 if self.coeff is None else self.coeff
+            coeff2 = 1 if other.coeff is None else other.coeff
+            return ScalarOp(self._input_dims, coeff=coeff1+coeff2)
+        return other._add(self).reshape(self._input_dims, self._output_dims)
+
+    def _multiply(self, other):
+        """Return the ScalarOp other * self.
+
+        Args:
+            other (complex): a complex number.
+
+        Returns:
+            ScalarOp: the scaled identity operator other * self.
+
+        Raises:
+            QiskitError: if other is not a valid complex number.
+        """
+        if not isinstance(other, Number):
+            raise QiskitError("other is not a number")
```

Comment:
this is exactly the type of thing type hints do, so you don't have to riddle your code with manual type checks

Reply:
I disagree. As far as I can tell Type hints are just hints, they don't stop you inserting any value and don't do error checking. eg: the following will execute without any warnings or errors

```
def foo(val: str) -> str:
    return val

foo([1,2,3])
```

Reply:
yes, python still remains a dynamically-typed language. However if the code has type hints other tools (such as MyPy) can catch programming errors when you use a disallowed type, without running the code. 

But my point here was you already say the expected type (in docstring in this case), so checking for it is a bit redundant unless there is a real pitfall of programmers passing in the wrong type.